File: taurus-protect-sdk-typescript/src/mappers/fiat.ts
/**
 * Fiat provider mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { TgvalidatordFiatProviderEntity } from '../internal/openapi/models/TgvalidatordFiatProviderEntity';
import type {
  FiatProvider,
  FiatProviderAccount,
  FiatProviderCounterpartyAccount,
  FiatProviderEntity,
  FiatProviderOperation,
} from '../models/fiat';
import { currencyFromDto } from './currency';
import { safeDate, safeMap, safeString } from './base';

/**
 * Maps a fiat provider DTO to a FiatProvider domain model.
 */
export function fiatProviderFromDto(dto: unknown): FiatProvider | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    provider: safeString(d.provider),
    label: safeString(d.label),
    baseCurrencyValuation: safeString(d.baseCurrencyValuation),
  };
}

/**
 * Maps an array of fiat provider DTOs to FiatProvider domain models.
 */
export function fiatProvidersFromDto(
  dtos: unknown[] | null | undefined
): FiatProvider[] {
  return safeMap(dtos, fiatProviderFromDto);
}

/**
 * Maps a fiat provider account DTO to a FiatProviderAccount domain model.
 */
export function fiatProviderAccountFromDto(
  dto: unknown
): FiatProviderAccount | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    provider: safeString(d.provider),
    label: safeString(d.label),
    accountType: safeString(d.accountType),
    accountIdentifier: safeString(d.accountIdentifier),
    accountName: safeString(d.accountName),
    totalBalance: safeString(d.totalBalance),
    currencyId: safeString(d.currencyID ?? d.currencyId),
    currencyInfo: d.currencyInfo ? currencyFromDto(d.currencyInfo) : undefined,
    baseCurrencyValuation: safeString(d.baseCurrencyValuation),
    creationDate: safeDate(d.creationDate),
    updateDate: safeDate(d.updateDate),
  };
}

/**
 * Maps an array of fiat provider account DTOs to FiatProviderAccount domain models.
 */
export function fiatProviderAccountsFromDto(
  dtos: unknown[] | null | undefined
): FiatProviderAccount[] {
  return safeMap(dtos, fiatProviderAccountFromDto);
}

/**
 * Maps a fiat provider counterparty account DTO to a FiatProviderCounterpartyAccount domain model.
 */
export function fiatProviderCounterpartyAccountFromDto(
  dto: unknown
): FiatProviderCounterpartyAccount | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    provider: safeString(d.provider),
    label: safeString(d.label),
    accountType: safeString(d.accountType),
    accountIdentifier: safeString(d.accountIdentifier),
    accountName: safeString(d.accountName),
    counterpartyId: safeString(d.counterpartyID ?? d.counterpartyId),
    counterpartyName: safeString(d.counterpartyName),
    currencyId: safeString(d.currencyID ?? d.currencyId),
    currencyInfo: d.currencyInfo ? currencyFromDto(d.currencyInfo) : undefined,
    creationDate: safeDate(d.creationDate),
    updateDate: safeDate(d.updateDate),
  };
}

/**
 * Maps an array of fiat provider counterparty account DTOs to domain models.
 */
export function fiatProviderCounterpartyAccountsFromDto(
  dtos: unknown[] | null | undefined
): FiatProviderCounterpartyAccount[] {
  return safeMap(dtos, fiatProviderCounterpartyAccountFromDto);
}

/**
 * Maps a fiat provider operation DTO to a FiatProviderOperation domain model.
 */
export function fiatProviderOperationFromDto(
  dto: unknown
): FiatProviderOperation | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    provider: safeString(d.provider),
    label: safeString(d.label),
    operationType: safeString(d.operationType),
    operationIdentifier: safeString(d.operationIdentifier),
    operationDirection: safeString(d.operationDirection),
    status: safeString(d.status),
    amount: safeString(d.amount),
    currencyId: safeString(d.currencyID ?? d.currencyId),
    currencyInfo: d.currencyInfo ? currencyFromDto(d.currencyInfo) : undefined,
    fromAccountId: safeString(d.fromAccountID ?? d.fromAccountId),
    toAccountId: safeString(d.toAccountID ?? d.toAccountId),
    fromDetails: safeString(d.fromDetails),
    toDetails: safeString(d.toDetails),
    comment: safeString(d.comment),
    operationDetails: safeString(d.operationDetails),
    creationDate: safeDate(d.creationDate),
    updateDate: safeDate(d.updateDate),
  };
}

/**
 * Maps an array of fiat provider operation DTOs to domain models.
 */
export function fiatProviderOperationsFromDto(
  dtos: unknown[] | null | undefined
): FiatProviderOperation[] {
  return safeMap(dtos, fiatProviderOperationFromDto);
}

/**
 * Maps a fiat provider entity row to a FiatProviderEntity domain model.
 */
export function fiatProviderEntityFromDto(
  dto: TgvalidatordFiatProviderEntity | null | undefined
): FiatProviderEntity | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }
  return {
    id: dto.id,
    provider: dto.provider,
    label: dto.label,
    accountIdentifier: dto.accountIdentifier,
    name: dto.name,
    details: dto.details,
    creationDate: dto.creationDate,
    updateDate: dto.updateDate,
  };
}

/**
 * Maps an array of fiat provider entity rows to FiatProviderEntity domain models.
 */
export function fiatProviderEntitiesFromDto(
  dtos: TgvalidatordFiatProviderEntity[] | null | undefined
): FiatProviderEntity[] {
  return safeMap(dtos, fiatProviderEntityFromDto);
}

File: taurus-protect-sdk-go/pkg/protect/mapper/fiat.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// FiatProviderFromDTO converts an OpenAPI TgvalidatordFiatProvider to a domain FiatProvider.
func FiatProviderFromDTO(dto *openapi.TgvalidatordFiatProvider) *model.FiatProvider {
	if dto == nil {
		return nil
	}

	return &model.FiatProvider{
		Provider:              safeString(dto.Provider),
		Label:                 safeString(dto.Label),
		BaseCurrencyValuation: safeString(dto.BaseCurrencyValuation),
	}
}

// FiatProvidersFromDTO converts a slice of OpenAPI TgvalidatordFiatProvider to domain FiatProviders.
func FiatProvidersFromDTO(dtos []openapi.TgvalidatordFiatProvider) []*model.FiatProvider {
	if dtos == nil {
		return nil
	}
	providers := make([]*model.FiatProvider, len(dtos))
	for i := range dtos {
		providers[i] = FiatProviderFromDTO(&dtos[i])
	}
	return providers
}

// FiatProviderAccountFromDTO converts an OpenAPI TgvalidatordFiatProviderAccount to a domain FiatProviderAccount.
func FiatProviderAccountFromDTO(dto *openapi.TgvalidatordFiatProviderAccount) *model.FiatProviderAccount {
	if dto == nil {
		return nil
	}

	account := &model.FiatProviderAccount{
		ID:                    safeString(dto.Id),
		Provider:              safeString(dto.Provider),
		Label:                 safeString(dto.Label),
		AccountType:           safeString(dto.AccountType),
		AccountIdentifier:     safeString(dto.AccountIdentifier),
		AccountName:           safeString(dto.AccountName),
		TotalBalance:          safeString(dto.TotalBalance),
		CurrencyID:            safeString(dto.CurrencyID),
		BaseCurrencyValuation: safeString(dto.BaseCurrencyValuation),
	}

	// Convert dates
	if dto.CreationDate != nil {
		account.CreationDate = *dto.CreationDate
	}
	if dto.UpdateDate != nil {
		account.UpdateDate = *dto.UpdateDate
	}

	// Convert currency info
	if dto.CurrencyInfo != nil {
		account.CurrencyInfo = CurrencyFromDTO(dto.CurrencyInfo)
	}

	return account
}

// FiatProviderAccountsFromDTO converts a slice of OpenAPI TgvalidatordFiatProviderAccount to domain FiatProviderAccounts.
func FiatProviderAccountsFromDTO(dtos []openapi.TgvalidatordFiatProviderAccount) []*model.FiatProviderAccount {
	if dtos == nil {
		return nil
	}
	accounts := make([]*model.FiatProviderAccount, len(dtos))
	for i := range dtos {
		accounts[i] = FiatProviderAccountFromDTO(&dtos[i])
	}
	return accounts
}

// FiatProviderEntityFromDTO converts an OpenAPI FiatProviderEntity to a domain FiatProviderEntity.
func FiatProviderEntityFromDTO(dto *openapi.TgvalidatordFiatProviderEntity) *model.FiatProviderEntity {
	if dto == nil {
		return nil
	}
	entity := &model.FiatProviderEntity{
		ID:                safeString(dto.Id),
		Provider:          safeString(dto.Provider),
		Label:             safeString(dto.Label),
		AccountIdentifier: safeString(dto.AccountIdentifier),
		Name:              safeString(dto.Name),
		Details:           safeString(dto.Details),
	}
	if dto.CreationDate != nil {
		entity.CreationDate = *dto.CreationDate
	}
	if dto.UpdateDate != nil {
		entity.UpdateDate = *dto.UpdateDate
	}
	return entity
}

// FiatProviderEntitiesFromDTO converts a slice of OpenAPI FiatProviderEntity to domain entities.
func FiatProviderEntitiesFromDTO(dtos []openapi.TgvalidatordFiatProviderEntity) []*model.FiatProviderEntity {
	if dtos == nil {
		return nil
	}
	entities := make([]*model.FiatProviderEntity, len(dtos))
	for i := range dtos {
		entities[i] = FiatProviderEntityFromDTO(&dtos[i])
	}
	return entities
}
